tests/unit/test_charm.py

```diff
@@ -0,0 +1,170 @@
+from ops import testing
+from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
+
+from charm import PrometheusConfigurerOperatorCharm
+
+testing.SIMULATE_CAN_CONNECT = True
+TEST_MULTITENANT_LABEL = "some_test_label"
+TEST_CONFIG = f"""options:
+  multitenant_label:
+    type: string
+    description: |
+      Prometheus Configurer has been designed to support multiple tenants. This label can be used
+      to restrict the alerting rules in Prometheus, so that each rule can only be triggered by
+      metrics with matching label.
+    default: {TEST_MULTITENANT_LABEL}
+"""
+
+
+class TestPrometheusConfigurerOperatorCharm(unittest.TestCase):
+    @patch(
+        "charm.KubernetesServicePatch",
+        lambda charm, ports: None,
+    )
+    def setUp(self):
+        self.harness = testing.Harness(PrometheusConfigurerOperatorCharm, config=TEST_CONFIG)
+        self.addCleanup(self.harness.cleanup)
+        self.harness.set_leader(True)
+        self.harness.begin()
+
+    @patch("charm.AlertRulesDirWatcher")
+    @patch("charm.PrometheusConfigurerOperatorCharm.RULES_DIR", new_callable=PropertyMock)
+    def test_given_rules_directory_when_pebble_ready_then_watchdog_starts_watching_given_rules_directory(  # noqa: E501
+        self, patched_rules_dir, patched_alert_rules_dir_watcher
+    ):
+        test_rules_dir = "/test/rules/dir"
+        patched_rules_dir.return_value = test_rules_dir
+        self.harness.container_pebble_ready("prometheus-configurer")
+
+        patched_alert_rules_dir_watcher.assert_called_with(self.harness.charm, test_rules_dir)
+
+    @patch("charm.AlertRulesDirWatcher", Mock())
+    def test_given_prometheus_relation_not_created_when_pebble_ready_then_charm_goes_to_blocked_state(  # noqa: E501
+        self,
+    ):
+        self.harness.container_pebble_ready("prometheus-configurer")
+
+        assert self.harness.charm.unit.status == BlockedStatus(
+            "Waiting for prometheus relation to be created"
+        )
+
+    @patch("charm.PrometheusConfigurerOperatorCharm.model")
+    @patch("charm.AlertRulesDirWatcher", Mock())
+    def test_given_prometheus_relation_created_but_prometheus_configurer_container_not_yet_ready_when_pebble_ready_then_charm_goes_to_waiting_state(  # noqa: E501
+        self, patched_relations_created
+    ):
+        patched_relations_created.return_value = True
+        testing.SIMULATE_CAN_CONNECT = False
+        self.harness.container_pebble_ready("prometheus-configurer")
+
+        assert self.harness.charm.unit.status == WaitingStatus(
+            "Waiting for prometheus-configurer container to be ready"
+        )
+
+    @patch("charm.PrometheusConfigurerOperatorCharm.RULES_DIR", new_callable=PropertyMock)
+    @patch(
+        "charm.PrometheusConfigurerOperatorCharm.PROMETHEUS_CONFIGURER_PORT",
+        new_callable=PropertyMock,
+    )
+    @patch(
+        "charm.PrometheusConfigurerOperatorCharm._prometheus_server_host",
```

Comment:
I don't think this needs to be patched, you can add configs using the self.harness.config or something like it

Reply:
The idea is that we can change the internal implementation of the class (Ex. change the property to another name) without the tests failing.

Reply:
Yes, I don't really know what happened here, because that's not even what I intended to patch... The idea was to patch `DUMMY_HTTP_SERVER_HOST` which should be returned by the `_prometheus_server_host` property. 
Anyway, I decided to remove both config option and the property. I don't expect this charm to be used with any other Prometheus than the `prometheus-k8s-operator` (which reloads automatically).